diff -Naur uboot/drivers/net/phy/broadcom.c uboot_new/drivers/net/phy/broadcom.c
--- uboot/drivers/net/phy/broadcom.c	2020-07-17 19:39:20.521367435 +0800
+++ uboot_new/drivers/net/phy/broadcom.c	2020-07-17 19:35:54.367733395 +0800
@@ -72,6 +72,12 @@
 		phy_write(phydev, MDIO_DEVAD_NONE, MIIM_BCM54XX_SHD, 0xc00);
 		mii_reg = 0x8c00;
 		phy_write(phydev, MDIO_DEVAD_NONE, MIIM_BCM54XX_SHD, mii_reg);		
+	} else if (phydev->drv->uid == 0x03625e6a) {
+		/* Disable RGMII RXD to RXC Skew */
+		phy_write(phydev, MDIO_DEVAD_NONE, 0x1c, 0x8c00);
+	
+		/* First Switch shadow register selector */
+		phy_write(phydev, MDIO_DEVAD_NONE, 0x18, 0xf0e7);
 	}
 
 	return 0;
@@ -348,6 +354,16 @@
 	.shutdown = &genphy_shutdown,
 };
 
+static struct phy_driver BCM54612_driver = {
+	.name = "Broadcom BCM54612",
+	.uid = 0x03625e6a,
+	.mask = 0xffffffff,
+	.features = PHY_GBIT_FEATURES,
+	.config = &bcm5461_config,
+	.startup = &bcm54xx_startup,
+	.shutdown = &genphy_shutdown,
+};
+
 static struct phy_driver BCM5461S_driver = {
 	.name = "Broadcom BCM5461S",
 	.uid = 0x2060c0,
@@ -391,6 +407,7 @@
 int phy_broadcom_init(void)
 {
 	phy_register(&BCM54616S_driver);
+	phy_register(&BCM54612_driver);
 	phy_register(&BCM5482S_driver);
 	phy_register(&BCM5464S_driver);
 	phy_register(&BCM5461S_driver);
