--- rsync-3.1.2.orig/options.c	2016-03-29 11:03:33.260097224 +0200
+++ rsync-3.1.2/options.c	2016-03-29 11:03:43.859729633 +0200
@@ -2121,6 +2121,7 @@
 	}
 	if (backup_dir) {
 		size_t len;
+		make_backups = 1; /* --backup-dir implies --backup */
 		while (*backup_dir == '.' && backup_dir[1] == '/')
 			backup_dir += 2;
 		if (*backup_dir == '.' && backup_dir[1] == '\0')
--- rsync-3.1.2.orig/rsync.yo	2015-12-30 11:52:56.599111311 +0100
+++ rsync-3.1.2/rsync.yo	2016-03-29 11:04:36.767894855 +0200
@@ -780,7 +780,7 @@
 your rules specify a trailing inclusion/exclusion of '*', the auto-added
 rule would never be reached).
 
-dit(bf(--backup-dir=DIR)) In combination with the bf(--backup) option, this
+dit(bf(--backup-dir=DIR)) This implies the bf(--backup) option, and
 tells rsync to store all backups in the specified directory on the receiving
 side.  This can be used for incremental backups.  You can additionally
 specify a backup suffix using the bf(--suffix) option
