--- rsync-3.1.2.orig/rsync.yo	2015-12-30 11:47:24.180646652 +0100
+++ rsync-3.1.2/rsync.yo	2015-12-30 11:47:27.477198899 +0100
@@ -1907,7 +1907,9 @@
 that will not be compressed.
 
 dit(bf(--compress-level=NUM)) Explicitly set the compression level to use
-(see bf(--compress)) instead of letting it default.  If NUM is non-zero,
+(see bf(--compress)) instead of letting it default.  Allowed values for NUM
+are between 0 and 9; default when bf(--compress) option is specified is 6.
+If NUM is non-zero,
 the bf(--compress) option is implied.
 
 dit(bf(--skip-compress=LIST)) Override the list of file suffixes that will
