diff -Naur linux_orig/drivers/mtd/parsers/spx-fmh.c linux/drivers/mtd/parsers/spx-fmh.c
--- linux_orig/drivers/mtd/parsers/spx-fmh.c	2021-04-09 19:36:41.508966831 +0530
+++ linux/drivers/mtd/parsers/spx-fmh.c	2021-04-09 19:51:48.661671707 +0530
@@ -23,6 +23,7 @@
 
 #ifdef CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT
 unsigned char broken_spi_banks = 0;// specify the number of broken SPI flash bank
+unsigned long spi_bank_size = 0;// specify the SPI flash bank size
 #endif
 
 
@@ -84,7 +85,7 @@
 #endif
         {
             //img2_offset+=ractrends_mtd[i]->size;
-            img2_offset+=CONFIG_SPX_FEATURE_GLOBAL_USED_FLASH_SIZE;
+            img2_offset+=FlashSize/2;
         }
     }
 #endif
@@ -118,6 +119,17 @@
 	//address computation for single image
 	endingsector=(unsigned long)div_u64(FlashSize,SectorSize);
 #endif	
+
+#ifdef CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT
+        if(broken_spi_banks == 1 || broken_spi_banks == 2)
+        {
+                spi_bank_size=FlashSize;
+        }
+        else
+        {
+                spi_bank_size=FlashSize/2;
+        }
+#endif
 	/* Create a MTD for full Flash */
 #if 0 	/* DT Kernel creates "mtd0"  for full flash */
 	parts[PartCount].name = kstrndup("fullpart",16,GFP_KERNEL);
@@ -236,10 +248,16 @@
 	}*/
 	if(0){
 #if defined (CONFIG_SPX_FEATURE_GLOBAL_DUAL_IMAGE_SUPPORT)	
+#ifdef CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT
+	if(strstr(boot_command_line,IMAGE_2)){
+		flashstart=(mtd->size/2);
+	}
+#else
 	if(strstr(boot_command_line,IMAGE_2)){
 		   flashstart=CONFIG_SPX_FEATURE_GLOBAL_USED_FLASH_SIZE;
 	}
 #endif
+#endif
 }
 	PartCount = GetFMHPartitions(mtd,flashstart,mtd->size,mtd->erasesize,tmp_parts);
 
@@ -338,6 +356,7 @@
 
 #ifdef CONFIG_SPX_FEATURE_HW_FAILSAFE_BOOT
 EXPORT_SYMBOL(broken_spi_banks) ;
+EXPORT_SYMBOL(spi_bank_size);
 #endif
 
 MODULE_AUTHOR("American Megatrends International");
