--- linux-5.4.41/Makefile	2020-05-14 01:58:30.000000000 -0400
+++ linux/Makefile	2020-05-17 03:15:27.849707192 -0400
@@ -2,7 +2,7 @@
 VERSION = 5
 PATCHLEVEL = 4
 SUBLEVEL = 41
-EXTRAVERSION =
+EXTRAVERSION = -ami
 NAME = Kleptomaniac Octopus
 
 # *DOCUMENTATION*
@@ -452,6 +452,7 @@
 		-I$(objtree)/arch/$(SRCARCH)/include/generated \
 		$(if $(building_out_of_srctree),-I$(srctree)/include) \
 		-I$(objtree)/include \
+		-include ${SPXINC}/projdef.h \
 		$(USERINCLUDE)
 
 KBUILD_AFLAGS   := -D__ASSEMBLY__ -fno-PIE
