diff -Naur linux/drivers/mtd/spi-nor/spi-nor.c linux.new/drivers/mtd/spi-nor/spi-nor.c
--- linux/drivers/mtd/spi-nor/spi-nor.c	2020-06-04 17:07:31.500321832 +0800
+++ linux.new/drivers/mtd/spi-nor/spi-nor.c	2020-06-04 17:18:05.813109308 +0800
@@ -2289,6 +2289,10 @@
 			SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
 	{ "is25wp512m", INFO( 0x9d601a, 0,64  * 1024, 1024,
 			SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
+	{ "is25lp512m",INFO(0x9d601a, 0,64  * 1024, 1024,
+			SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
+	{ "is25lp256d", INFO( 0x9d6019, 0, 64 * 1024, 512,
+			SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
 
 	/* Macronix */
 	{ "mx25l512e",   INFO(0xc22010, 0, 64 * 1024,   1, SECT_4K) },
