diff -Naur linux_org/drivers/net/ethernet/faraday/ftgmac100.c linux/drivers/net/ethernet/faraday/ftgmac100.c
--- linux_org/drivers/net/ethernet/faraday/ftgmac100.c	2021-04-15 10:48:09.678454847 +0800
+++ linux/drivers/net/ethernet/faraday/ftgmac100.c	2021-04-22 14:59:01.350390294 +0800
@@ -52,6 +52,8 @@
 #define FTGMAC_100MHZ		100000000
 #define FTGMAC_25MHZ		25000000
 
+long unsigned int ftgmac100_mac_register[4] = {0x1e660000, 0x1e680000, 0x1e670000, 0x1e690000};
+
 struct ftgmac100 {
 	/* Registers */
 	struct resource *res;
@@ -1748,7 +1750,8 @@
 	struct net_device *netdev;
 	struct ftgmac100 *priv;
 	struct device_node *np;
-	int err = 0;
+	int err = 0, index = 0;
+	long unsigned int register_addr = 0;
 
 	if (!pdev)
 		return -ENODEV;
@@ -1881,12 +1884,19 @@
 		phy_attached_info(phy);
 	}else{
 		priv->use_ncsi = false;
-		err = ftgmac100_setup_mdio(netdev);
-		if (err)
-		{
+#ifdef CONFIG_SPX_FEATURE_NCSI_DEFAULT_INTERFACE
+		sscanf(CONFIG_SPX_FEATURE_NCSI_DEFAULT_INTERFACE, "eth%d", &index);
+		sscanf(pdev->name, "%lx.ethernet", &register_addr);
+		if (ftgmac100_mac_register[index] == register_addr) {
 			//Enable interface for NCSI usage
 			dev_info(&pdev->dev, "Using NCSI interface\n");
 			priv->use_ncsi = true;
+		}else 
+#endif
+		{
+			err = ftgmac100_setup_mdio(netdev);
+			if (err)
+				goto err_setup_mdio;
 		}
 	}
 
