From: Enrico Tassi <gareuselesinge@debian.org>
Date: Fri, 9 Aug 2013 17:47:14 +0200
Subject: lua-sec.h

---
 lua-sec.h | 9 +++++++++
 1 file changed, 9 insertions(+)
 create mode 100644 lua-sec.h

diff --git a/lua-sec.h b/lua-sec.h
new file mode 100644
index 0000000..35281bf
--- /dev/null
+++ b/lua-sec.h
@@ -0,0 +1,9 @@
+/* This file has been added in the Debian package */
+
+extern int luaopen_ssl_core(lua_State*);
+extern int luaopen_ssl_context(lua_State*);
+extern int luaopen_ssl_x509(lua_State*);
+extern int luaopen_ssl_ec(lua_State*);
+extern int luaopen_ssl_config(lua_State*);
+
+// eof
